src/transformers/models/clvp/modeling_clvp.py

```diff
@@ -0,0 +1,1954 @@
+    @replace_return_docstrings(output_type=CLVPOutput, config_class=CLVPConfig)
+    def forward(
+        self,
+        input_ids: torch.LongTensor = None,
+        input_features: torch.FloatTensor = None,
+        attention_mask: Optional[torch.LongTensor] = None,
+        use_causal_attention_mask: Optional[bool] = False,
+        return_loss: Optional[bool] = None,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+    ) -> Union[Tuple, CLVPOutput]:
+        r"""
+        Returns:
+
+        Examples:
+
+        ```python
+        >>> import datasets
+        >>> from transformers import CLVPProcessor, CLVPModel
+
+        >>> # Define the Text and Load the Audio (We are taking an audio example from HuggingFace Hub using `datasets` library)
+        >>> text = "This is an example text."
+
+        >>> ds = datasets.load_dataset("hf-internal-testing/librispeech_asr_dummy", "clean", split="validation")
+        >>> ds = ds.cast_column("audio", datasets.Audio(sampling_rate=22050))
+        >>> _, audio, sr = ds.sort("id").select(range(1))[:1]["audio"][0].values()
+
+        >>> # Define processor and model
+        >>> processor = CLVPProcessor.from_pretrained("susnato/clvp_dev")
+        >>> model = CLVPModel.from_pretrained("susnato/clvp_dev")
+
+        >>> # processor outputs and model output s
+        >>> processor_output = processor(raw_speech=audio, sampling_rate=sr, text=text, return_tensors="pt")
+        >>> outputs = model(
+        ...     input_ids=processor_output["input_ids"],
+        ...     input_features=processor_output["input_features"],
+        ...     return_dict=True,
+        ... )
+        ```
+        """
+
+        # Use CLVP model's config for some fields (if specified) instead of those of speech & text components.
+        output_attentions = output_attentions if output_attentions is not None else self.config.output_attentions
+        output_hidden_states = (
+            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
+        )
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        conditioning_embeds = self.conditioning_encoder(mel_spec=input_features, text_tokens=input_ids)
+
+        speech_candidates = self.autoregressive_model(
+            inputs_embeds=conditioning_embeds,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+        )
+
+        speech_candidates = speech_candidates[0]
+
+        # since we will get the embeds of shape `(batch_size, seq_len, embedding_dim)` during the forward pass
+        # we must convert it to tokens, to make it compaitable with speech_transformer
+        if speech_candidates.ndim == 3:
+            speech_candidates = speech_candidates.argmax(2)
```

Comment:
Hi @ylacombe , I defined the forward pass here because as you said [here](https://github.com/huggingface/transformers/pull/24745#issuecomment-1678578741), not to mix up the forward and generate, so we use the forward of the autoreg model during the forward pass of the `CLVPModel`.  

Is it ok if I don't define the forward pass for the `CLVPModel` and only keep `generate`? Since in theory the speech encoder should only be used on the speech candidates generated by the autoreg model and it does not make any sense to use the speech model to use on the forward pass of the autoreg model.

Reply:
cc @ArthurZucker @amyeroberts

This is something that Arthur and Amy might help us decide.

To give a little bit more context, `CLPVModel` is a CLIP-like model for audio. It is made of:

- GPT2-like model under-the-hood to generate audio tokens
- 2 autoencoder that shares the same architecture, one for audio and one for text.

Since it needs to `generate` audio tokens, the model don't have a straightforward forward method.

`Our question:` Is it worth to have a forward pass?

A solution to keep a forward method could be to make `forward` compatible with the text autoencoding (i.e use text token ids and the text autoencoder). Whereas the generate method would take care of both.
WDYT of that ?

Thanks for your help !